Sources/System/Sockets/SocketDescriptor+Messages.swift

```diff
@@ -0,0 +1,333 @@
+        let buffer = UnsafeRawBufferPointer(rebasing: buffer.prefix(_endOffset))
+        return try body(buffer)
+      }
+    }
+
+    internal mutating func _withUnsafeMutableBytes<R>(
+      _ body: (UnsafeMutableRawBufferPointer) throws -> R
+    ) rethrows -> R {
+      return try _buffer.withUnsafeMutableBytes { buffer in
+        assert(buffer.count >= _endOffset)
+        let buffer = UnsafeMutableRawBufferPointer(rebasing: buffer.prefix(_endOffset))
+        return try body(buffer)
+      }
+    }
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor.ControlMessageBuffer: Collection {
+  @frozen
+  public struct Index: Comparable, Hashable {
+    @usableFromInline
+    var _offset: Int
+
+    @inlinable
+    internal init(_offset: Int) {
+      self._offset = _offset
+    }
+
+    @inlinable
+    public static func == (left: Self, right: Self) -> Bool {
+      left._offset == right._offset
+    }
+
+    @inlinable
+    public static func < (left: Self, right: Self) -> Bool {
+      left._offset < right._offset
+    }
+
+    @inlinable
+    public func hash(into hasher: inout Hasher) {
+      hasher.combine(_offset)
+    }
+  }
+
+  public struct Message {
+    internal var _base: SocketDescriptor.ControlMessageBuffer
+    internal var _offset: Int
+
+    internal init(_base: SocketDescriptor.ControlMessageBuffer, offset: Int) {
+      self._base = _base
+      self._offset = offset
+    }
+  }
+
+  public var startIndex: Index { Index(_offset: 0) }
+  public var endIndex: Index { Index(_offset: _endOffset) }
+  public var isEmpty: Bool { _endOffset == 0 }
+
+  public var count: Int {
+    _withUnsafeBytes { buffer in
+      var count = 0
+      var p = buffer.baseAddress!
+      let end = buffer.baseAddress! + buffer.count
+      while p <= end - _headerSize {
+        count += 1
+        let header = p.assumingMemoryBound(to: CInterop.CMsgHdr.self)
+        let length = Int(header.pointee.cmsg_len)
+        p += length
+      }
+      return count
+    }
+  }
+
+  public func index(after i: Index) -> Index {
+    precondition(i._offset != _endOffset, "Can't advance past endIndex")
+    precondition(i._offset >= 0 && i._offset + _headerSize <= _endOffset,
+                 "Invalid index")
+    let delta: Int = _withUnsafeBytes { buffer in
+      let p = (buffer.baseAddress! + i._offset)
+        .assumingMemoryBound(to: CInterop.CMsgHdr.self)
+      return Int(p.pointee.cmsg_len)
+    }
+    if i._offset + delta + _headerSize > _endOffset {
+      return endIndex
+    }
+    return Index(_offset: i._offset + delta)
+  }
+
+  public subscript(position: Index) -> Message {
+    precondition(
+      position._offset >= 0 && position._offset + _headerSize <= _endOffset,
+      "Invalid index")
+    return Element(_base: self, offset: position._offset)
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor.ControlMessageBuffer.Message {
```

Comment:
I got a little lost. Why is `Message` nested inside  of `ControlMessageBuffer`? Isn't this actually the header?

Naively, I would have thought to have 

```
SocketDescriptor
  - MessageHeader
    - Name: SocketAddress
    - Data: Raw iov or alternative
    - AncillaryData: [ControlMessage]
    - Flags
```

Or maybe I'm mixing up at what level there is a "message" here...

Reply:
`Message` is an individual message inside `ControlMessageBuffer` (now called `AncillaryMessageBuffer`); it's the element type of the collection.

I got rid of `MessageHeader` -- we didn't really need it. (The drawback is that we have a multitude of `receiveMessage` overloads.)

Reply:
So `sendmsg`/`recvmsg` operate on the following data:

- A socket address representing the remote party (`msg_name`)
- An IO vector holding the transferred bytes (`msg_iov`)
- A sequence of an arbitrary number of variable-length control/ancillary messages, encoded in a single contiguous buffer (`msg_control`)
- Message flags describing the received message (only recvmsg) (`msg_flags`)

AncillaryMessageBuffer corresponds to `msg_control`. (This has nothing to do with the actual message transmitted -- that's always stored in the iov.)